ls1intum/Artemis
# [Feature] Update Template on Import

### Checklist

### Motivation and Context


This PR adds an option to the import of programming exercises which updates the template files with the latest template of Artemis. In later iterations of this feature, this will allow instructors to enable/disable feature which were locked on import before. For this first iteration the Maven pom files found in the test, exercise, solution repositories are updated with dependencies and plugins found in the latest template. This only works for non-sequential tests for now.

### Description

- Implements a strategy pattern for different upgrade strategies per programming language
- Implements a strategy for Java which updates the repository project object models plugins and dependencies with the ones found in the template.

### Steps for Testing


1. Log in to Artemis
2. Identify a old Java programming exercise (still using JUnit4) or manipulate pom files of Java exercises yourself (add plugins, dependencies, change versions and JDK, add JUNIT
3. Import this exercise, and activate the Upgrade Template option
4. Check the pom.xml files in the Test/Solution/Template repository using 'Edit in Editor'
--> JUnit should be gone
--> Legacy dependencies org.json.json and me.xdrop.fuzzywuzzy should be gone
--> AJTS should be added
--> Plugins failsafe, surefire, compiler should have the same version and configuration as in the latest template (check the JDK version)
--> Overall, all dependencies except ones the instructor added himself should use the version of found in the template.
**The tests will most likely not work out of the box after activating this option. This is expected**
5. Also test this with a maven project type programming exercise. As they are quite new, you will have to fabricate the pom.xml files for this exercise yourself e.g. downgrade to JDK 14, introduce junit4...
6. Quickly create a programming exercise with SCA and another one with sequential test runs to check that this still works.


### Test Coverage



Decreases. The new functionality file be covered in a follow up.

### Screenshots



![image](https://user-images.githubusercontent.com/16407766/99850223-45fc4880-2b7d-11eb-8781-0ed68207b9e7.png)
